src/cargo/util/paths.rs

```diff
@@ -180,6 +180,42 @@ pub fn mtime(path: &Path) -> CargoResult<FileTime> {
     Ok(FileTime::from_last_modification_time(&meta))
 }
 
+/// Returns the maximum mtime of the given path, recursing into
+/// subdirectories, and following symlinks.
+pub fn mtime_recursive(path: &Path) -> CargoResult<FileTime> {
+    let meta = fs::metadata(path).chain_err(|| format!("failed to stat `{}`", path.display()))?;
+    if !meta.is_dir() {
+        return Ok(FileTime::from_last_modification_time(&meta));
+    }
+    let max_meta = walkdir::WalkDir::new(path)
+        .follow_links(true)
+        .into_iter()
+        .filter_map(|e| e.ok())
```

Comment:
I'm a little nervous about ignoring errors here and below in `unwrap_or_else`. I suppose though that the thinking is that if Cargo can't read the file then the build script probably also couldn't read the file so it couldn't be considered input? I think it'd be good in any case to add some comments here about why errors are ignored.

Reply:
Correct, my thinking was that if Cargo can't access it, the build script can't either.

I pushed a commit with more comments, and added debug logging just in case.  It makes the code a little longer than I'd like, but such is life with good error handling.